### PR TITLE
63 visual mapping highlight those output nodes not being assigned 2

### PR DESCRIPTION
**Pull request by johannesheucher-gip**

### CODE REVIEW

File: document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts [hunk 1]
```diff
@@ -143,14 +143,19 @@ export class SkeletonTreeNode implements GraphicallyRepresented<Element>, Dragga
...
+
+
+    unmarkRecursively() {
+        this.markRecursively(false);
     }
 
 
     get allChildrenMarked(): boolean {
-        return !this.children.some(child => !child.marked);
+        return this.children.length > 0 && !this.children.some(child => !child.marked);
```
Comment (TorbenSiegismund-GIP) [inline document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts:158]: How can the information  propagate towards root?
Should the marked state change?
In my test is the a complex Type containing a complex object with allChildrenMarked = true not marked, but it should be.

File: document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts [hunk 2]
```diff
@@ -550,7 +561,10 @@ export class SkeletonTreeDataSource implements TreeNodeFactory, TreeNodeObserver
      * Clears all `marked` states of all nodes
      */
     clearMarks() {
-        this.root$.pipe(first()).subscribe(root => root.unmarkRecursively());
+        this.root$.pipe(
+            filter(root => !!root),
+            first()
```
Comment (TorbenSiegismund-GIP) [inline document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts:566]: Better 
this.root$.pipe(
            first()
            filter(root => !!root),

File: document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts [hunk 3]
```diff
@@ -154,8 +159,14 @@ export class SkeletonTreeNode implements GraphicallyRepresented<Element>, Dragga
...
 
 
-    get allChildrenMarked(): boolean {
-        return this.children.length > 0 && !this.children.some(child => !child.marked);
+    markIfChildrenMarked() {
+        // all children marked
+        if (this.children.length > 0 && !this.children.some(child => !child.marked)) {
+            this.marked = true;
+        }
+        if (!this.parent?.marked) {
```
Comment (TorbenSiegismund-GIP) [inline document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts:167]: Is it not enough to check parent, if (this.children.length > 0 && !this.children.some(child => !child.marked)) = true?

File: document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts (hunk 2 above)
Comment (johannesheucher-gip) [reply document/workflow/variable-tree/data-source/skeleton-tree-data-source.ts:566]: In your case, the subscription wouldn't be called for `root` being *null* at first. What I want is the first emit of `root` being not *null*.
I saw you can put a predicate inside of `first`. So `pipe(first(root => !!root))` is even better.

**Review by TorbenSiegismund-GIP — APPROVED** [review]

**Pull request merged**